docs/proposals/deploy.md

```diff
@@ -0,0 +1,152 @@
+  - [Related `kubectl` Commands](#related-kubectl-commands)
+    - [`kubectl run`](#kubectl-run)
+    - [`kubectl scale` and `kubectl autoscale`](#kubectl-scale-and-kubectl-autoscale)
+    - [`kubectl rollout`](#kubectl-rollout)
+    - [`kubectl set`](#kubectl-set)
+    - [Example](#example)
+  - [Support in Deployment](#support-in-deployment)
+    - [Deployment Status](#deployment-status)
+    - [Deployment Version](#deployment-version)
+    - [Inert Deployments](#inert-deployments)
+    - [Perm-failed Deployments](#perm-failed-deployments)
+
+<!-- END MUNGE: GENERATED_TOC -->
+
+# Deploy through CLI
+
+## Motivation
+
+Users can use Deployments or `kubectl rolling-update` to deploy in their Kubernetes clusters. A Deployment provides declarative update for Pods and ReplicationControllers, whereas `rolling-update` allows the users to update their earlier deployment without worrying about schemas and configurations. Users need a way that's similar to `rolling-update` to manage their Deployments more easily.
+
+`rolling-update` expects ReplicationController as the only resource type it deals with. It's not trivial to support exactly the same behavior with Deployment, which requires:
+- Print out scaling up/down events.
+- Stop the deployment if users press Ctrl-c.
+- The controller should not make any more changes once the process ends. (Delete the deployment when status.replicas=status.updatedReplicas=spec.replicas)
+
+So, instead, this document proposes another way to support easier deployment management via Kubernetes CLI (`kubectl`).
+
+## Requirements
+
+The followings are operations we need to support for the users to easily managing deployments:
+
+- **Create**: To create deployments.
+- **Rollback**: To restore to an earlier version of deployment.
+- **Watch the status**: To watch for the status update of deployments.
+- **Pause/resume**: To pause a deployment mid-way, and to resume it. (A use case is to support canary deployment.)
+- **Version information**: To record and show version information that's meaningful to users. This can be useful for rollback.
+
+## Related `kubectl` Commands
+
+### `kubectl run`
+
+`kubectl run` should support the creation of Deployment (already implemented) and DaemonSet resources.
+
+### `kubectl scale` and `kubectl autoscale`
+
+Users may use `kubectl scale` or `kubectl autoscale` to scale up and down Deployments (both already implemented).
+
+### `kubectl rollout`
+
+`kubectl rollout` supports both Deployment and DaemonSet. It has the following subcommands:
+- `kubectl rollout undo` works like rollback; it allows the users to rollback to a previous version of deployment.
+- `kubectl rollout pause` allows the users to pause a deployment.
+- `kubectl rollout resume` allows the users to resume a paused deployment.
+- `kubectl rollout status` shows the status of a deployment.
+- `kubectl rollout history` shows meaningful version information of all previous deployments.
+
+### `kubectl set`
```

Comment:
We had a fairly deep UX discussion today on commands that support targeted use case functionality for edit.  The root issue was `expose` - expose tries to be polymorphic to objects that have no common interface (pod -> service -> ingress / route).  Therefore, each flag we add to expose causes severe issues in help and other supporting behavior because "--port" means something different to each object.  

Our plan was to recommend we stop adding flags to expose (and rollback anything not necessary to have minimal expose work correctly), and instead invest in targeted edit.  I.e., expose gets you 80% of the way, but if you need deeper functionality you don't rely on more complex expose, you rely on a command that is targeted to helping you edit a particular use case.  Example is `kubectl set service-ports` or `kubectl set route-termination` or `kubectl set ingress-termination`.  A user, using the help function on set, would get a large list of possible things to edit, but each item in that list would be able to solve a particular use case.

The guiding principle would be:

    kubectl set <noun>-<subnoun>

where in a few cases, such as common use, or common interface, `<noun>` could be omitted.  As an example, pod manipulation is very common, and as noted below we expect many objects to have single primary pod templates (they are the core object of the system).  So for concepts that exist only on pods, the default noun would be omitted so `kubectl set env` and that command would work across any resource that is polymorphic to a pod template (imagine an interface for `PodTemplate() api.PodTemplate`).  Other types would have the `noun` in place when the object the noun acts on is very clear (ingress, daemonset, etc).

Root level commands would target polymorphic interfaces across large subsets of commands (scale, delete, get) or creation flows (create, run, expose) or very important actions (exec, portforward)